backend/routes/graph.py

```diff
@@ -29,166 +35,308 @@ def subgraph():
+    missing_nodes = []  # nodes needed to traverse all paths but not present in nodes list
+    for root in roots:
+        for leaf in leaves:
+            _paths = list(nx.all_simple_paths(G, root, leaf))
+            for path in _paths:
+                for node in path:
+                    if node not in nodes:
+                        missing_nodes.append(node)
+    sgc = G.subgraph(nodes + missing_nodes)
     return sgc
 
 
-def sort_by_most_descendants(G, nodes):
-    key = lambda n: len(nx.descendants(G, n))
-    return sorted(nodes, key=key, reverse=True)
+def paths_as_indented_tree(paths: List[List[int]]) -> List[Tuple[int, int]]:
+    """
+    paths: A list of paths (each path is a list of node identifiers).
+    Initializes a root TreeNode with None value.
+    Inserts each path into this tree.
+    Generates a list of lines where each line represents a level and a node identifier in the indented tree.
+    """
+    # Initialize tree and insert paths
+    root = TreeNode(None)  # Root node doesn't hold any data
+    # paths = [("A", "B", "C"), ("A", "B", "D"), ("E", "B", "C"), ("E", "B", "D")]
+    for path in paths:
+        root.insert(path)
 
-def graph_to_json(G, sorted=False):
-    root_nodes = [node for node in G.nodes if G.in_degree(node) == 0]
-    if sorted:
-        root_nodes = sort_by_most_descendants(G, root_nodes)
-    hierarchies = {root: create_hierarchy(G, root, sorted) for root in root_nodes}
-    return hierarchies
+    tree = []
+    for line in generate_csv_lines(root):
+        # tree.append({'level': line[0], 'concept_id': line[1]})
+        tree.append((line[0], line[1]))
 
+    return tree
 
-def create_hierarchy(G, node, sorted=False):
-    successors = list(G.successors(node))
-    if len(successors) == 0:
-        return None
-    if sorted:
-        successors = sort_by_most_descendants(successors)
-    else:
-        return {successor: create_hierarchy(G, successor, sorted) for successor in successors}
-
-def for_testing():
-    G = nx.DiGraph()
-    G.add_edges_from([(1, 2), (1, 3), (2, 4), (2, 5), (2,3), (2, 6), (3, 6), (9,2)])
-    return G
 
+class TreeNode:
+    """TODO: document how this works, what it's for (got it from chatgpt)"""
+    """
+    Purpose: Represents a node in the tree, used to construct the indented tree structure.
+    Attributes:
+    value: The value (or identifier) of the node.
+    children: A dictionary of child nodes.
+    Methods:
+    __init__(self, value): Constructor to initialize the node with a value.
+    insert(self, path): Inserts a path into the tree, starting from this node.
+    """
+    def __init__(self, value):
+        self.value = value
+        self.children = {}
+
+    def insert(self, path):
+        if path:
+            head, *tail = path
+            if head not in self.children:
+                self.children[head] = TreeNode(head)
+            self.children[head].insert(tail)
+
+
+def condense_super_nodes(sg, threshhold=10):
```

Comment:
I don't see where `condense_super_nodes()` and `expand_super_node()` are used yet but I guess that will be soon!